src/transport.rs

```diff
@@ -871,56 +936,37 @@ impl Connection {
         Ok(())
     }
 
-    fn write(&self, data: &str) -> Result<(), Error> {
-        let mut writer = self.writer.lock()?;
-        writer.write_all(data.as_bytes())?;
-        Ok(())
-    }
-
     fn write_message(&self, message: &RequestMessage) -> Result<(), Error> {
-        let mut writer = self.writer.lock()?;
-
-        let data = message.encode();
-        debug!("-> {data:?}");
-
-        let data = data.as_bytes();
-
-        let mut packet = Vec::with_capacity(data.len() + 4);
-
-        packet.write_u32::<BigEndian>(data.len() as u32)?;
-        packet.write_all(data)?;
-
-        writer.write_all(&packet)?;
-
         self.recorder.record_request(message);
-
+        let encoded = message.encode();
+        debug!("-> {encoded:?}");
+        let length_encoded = encode_length(&encoded);
```

Comment:
Nit: `encoded_with_length` might work better here. There is some ambiguity now between this being the length of the encoded message or the message encoded with length.